src/libstore/lock.cc

```diff
@@ -94,10 +94,15 @@ struct SimpleUserLock : UserLock
                 if (err == -1)
                     throw Error("failed to get list of supplementary groups for '%s'", pw->pw_name);
 
+                // If we didn't call `getgrouplist` twice, the vector is still
+                // too large. If we did, then this doesn't change anything.
+                gids.resize(ngroups);
+
```

Comment:
It would be nice to factor the construction of `gids` into a function.

Reply:
@tfc I finally get it, the problem is this `ngroups` variable is poor taste --- the only state we need is from the vector itself, and it should be gotten rid of.

```c++
                /* Get the list of supplementary groups of this build
                   user.  This is usually either empty or contains a
                   group such as "kvm".  */
                std::vector<gid_t> gids;
                gids.resize(32); // arbitrary initial guess
                
                int err;
                {
                    int ngroups;
                    err = getgrouplist(
                        pw->pw_name, pw->pw_gid,
                        gids.data(),
                        &ngroups);
                    // resize to be the right size;
                    gids.resize(ngroups);
                }
                
                if (err = -1) {
                    // Our initial size of wasn't sufficient, we know are the right size, so we try again.
                    err = getgrouplist(
                        pw->pw_name, pw->pw_gid,
                        gids.data(),
                        NULL);
                    // If it failed once more, then something must be broken.
                    if (err == -1)
                        throw Error("failed to get list of supplementary groups for '%s'", pw->pw_name);
                }
```

Reply:
if we cannot do `NULL` for `ngroups`, we should do a second `int ngroups` variable. This makes clear it's just an out parameter and not in in-out parameter.

Reply:
I refactored the code a bit, i hope you find it more comprehensible this way.

Reply:
Yeah can just follow up in the other thread.